Add overloading experiment blog post (#1907)
rust-lang/blog.rust-lang.org: content/inside-rust/overloading-experiment.md

+++

> path = "inside-rust/2026/08/10/overloading-experiment"

Author: ```suggestion
path = "inside-rust/2026/08/18/overloading-experiment"
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/blog.rust-lang.org/1907/cf62f850931521a105ca11bad51b4cc248b9bb7b..4c1227d02eda02313145036bc8f3c4aeac9874b6)*

title = "Rust Function Overloading - Call for Experimentation"

> authors = ["teor"]

Author: Since this is a lang experiment, do we want to put it under the lang team?

```suggestion
authors = ["teor"]

[extra]
team = "The Language Team"
team_url = "https://www.rust-lang.org/governance/teams/lang"
```

example [is available on GitHub][overloading-examples]. You can also run a minimal Rust-only
example online, [in the Rust playground][playground-hypot].

[Nadrieril’s original "Overloading at Home" code][pre-pre-rfc] from his recent write-up of "splat",
can also be run [in the Rust playground][playground-at-home].

[cpp-crate]: https://crates.io/crates/cpp
[overloading-examples]: https://github.com/rustfoundation/overloading-examples
[playground-hypot]: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=d9494d84fa32271b3517fa293d3215ff
[pre-pre-rfc]: https://internals.rust-lang.org/t/pre-pre-rfc-splatting-for-named-arguments-and-function-overloading/24012
[playground-at-home]: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=fcf1ead0e431166177fd9f05d5e33fd0

### Limitations

As stated above, splat is currently an incomplete compiler feature, and is only available in the
nightly Rust compiler. Splat is also unergonomic, and we want to change that as part of the next
design phase.

Splat isn’t currently supported [on function pointers][fnptr-bug], but hopefully

> [that implementation][fnptr-pr] will be merged soon. If you want overloading support for function

Author: The FnPtr implementation PR [has merged](https://github.com/rust-lang/rust/pull/159643), so this will need to change to:
```suggestion
We've just merged [splat support for function pointers][fnptr-pr], if you're seeing
[an internal compiler error][fnptr-bug], please update to the latest nightly.
If you want overloading support for function
```

It is driven by the [Nightly support for function overloading in FFI bindings][overloading-goal]
Rust Project goal, and it is an outcome of the
[C++/Rust Interop Problem Space Mapping][interop-goal] Rust Project goal.

[interop-goal]: https://rust-lang.github.io/rust-project-goals/2026/interop-problem-map.html

Thank you also to everyone who has contributed to the overloading work, including [Oli][gh-oli],
[Ajay][gh-ajay], [Nadrieril][gh-nadri], [Scott][gh-scott], [Taylor][gh-taylor], [Tyler][gh-tyler],
[Matthias][gh-matthias], [Tim][gh-tim], [Devin][gh-devin], [Ralf][gh-ralf], [Jacob][gh-jacob],
[Zachary][gh-zachary] and the many project members who have given suggestions, feedback, testing,
bug reports, and reviews.

[gh-oli]: https://github.com/oli-obk
[gh-nadri]: https://github.com/nadrieril
[gh-scott]: https://github.com/scottmcm
[gh-taylor]: https://github.com/cramertj
[gh-tyler]: https://github.com/tmandry
[gh-matthias]: https://github.com/matthiaskrgr
[gh-tim]: https://github.com/theemathas
[gh-devin]: https://github.com/ssbr
[gh-ralf]: https://github.com/RalfJung
[gh-jacob]: https://github.com/programmerjake
[gh-zachary]: https://github.com/bushrat011899

## Future Work

> ### Overloading, In The Shiny Future

Reviewer: It might be worth noting here that this work is currently targetted at interop, and so while this might be desirable for pure Rust code eventually. I assume that we might want to limit this to `extern` blocks for interop purposes, and that extending this to the language as a whole would be the separate, future endeavour.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/blog.rust-lang.org/1907/cf62f850931521a105ca11bad51b4cc248b9bb7b..3ea344373403c802f93ae3844c29edc3bdced2a1)*

Author: Thanks for the feedback! Was something like this what you had in mind?
https://github.com/rust-lang/blog.rust-lang.org/pull/1907/changes#r3778944129

```rust
#[overload]
impl f64 {
    /// Returns the 2-dimensional distance from the origin.
    fn hypot(self, y: f64) -> f64 { … }

    /// Returns the 3-dimensional distance from the origin.
    fn hypot(self, y: f64, z: f64) -> f64 { … }
}
```

_(This example is a variant of a design shared by [Taylor Cramer][gh-taylor].)_

In the shiny future, Rust might have an `#[overload]` attribute that "just works" to call
overloaded foreign functions with the same name. No traits, tuples, or `#[rustc_splat]` required,

> the compiler handles it all for you.

Author: ```suggestion
the compiler handles it all for you. Since this work is targeted at interop, overloading might be
limited to `extern` blocks to start with. Extending it to native Rust could be a separate feature,
stabilised on a longer timeframe (or not at all).
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/blog.rust-lang.org/1907/cf62f850931521a105ca11bad51b4cc248b9bb7b..3ea344373403c802f93ae3844c29edc3bdced2a1)*
